texastribune/dot
# Add Auth0 API authorization

#### What's this PR do?
Adds the ability to allow certain GraphQL queries and mutations only to be accessible to certain staff members. This is done by verifying the permissions on access tokens granted by Auth0 and registering the Dot API inside our Auth0 tenants.

Also improves error handling.

#### Why are we doing this? How does it help us?
The API behind this app should only be callable by staff members, for obvious reasons. And having this very explicit `dot:view_data` permission inside Auth0 means we can limit access even more granularly.

#### How should this be manually tested?
Add the following to your `env-docker`:
```
DB_HOST=dot-db
DB_USER=postgres
DB_PASSWORD=postgres
DB_NAME=dot
DB_PORT=5432
NODE_PORT=3000
APP_URL=http://localhost:3000
AUTH0_DOMAIN=auth-test.texastribune.org
AUTH0_API_AUDIENCE=https://texastribune.org/dot
AUTH0_CLIENT_ID=YQ7ktxaQjg0SJ4m2yMylVdF1N7KSgoqg
AUTH0_CLIENT_SECRET=[get from Auth0 test tenant under Applications --> Dot]
```
Then:
1. `make -f Makefile.dev`
2. `npm start`
3. Go to `http://localhost:3000/log-in/`. You'll be asked to log in and/or authorize access to the Dot app. Do that!
4. You should get redirected to another page. Open your browser console and unfold the object logged there until you find your access token (`data.tokens.access_token`). Copy it (don't include the beginning and end double quote).
5. Throw your token in [this website](https://jwt.io/) and confirm the `permissions` key of the decoded token is an empty array.
6. Execute the below cURL command in a new terminal window, replacing `<access token>` with yours. You should get back some `foo bar` JSON, but the terminal you're running `npm start` in should have a message about you not being allowed there.
7. In the Auth0 dashboard, inside the **test tenant**, go to Users & Roles --> Users (on the left). Find yourself, go to the Permissions tab, click Assign Permissions, click Dot API, then give yourself `dot:view_data` permissions.
8. Repeat steps 3-6. In step 5, the permissions array should now contain `dot:view_data`. In step 6, the message logged to your `npm start` console should say that you **are** allowed there.

```
curl \
  -X POST \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer <access token>" \
  --data '{ "query": "{ posts }" }' \
  http://localhost:3000/graph/
```

#### How should this change be communicated to end users?
NA.

#### Are there any smells or added technical debt to note?
Nope! I think this is the clean, right way to do API authorization with Auth0. 

#### What are the relevant tickets?
https://3.basecamp.com/3098728/buckets/736178/todos/2545308227

#### Have you done the following, if applicable:
